### PR TITLE
Utilize `perf_context`

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

rust-rocksdb exposed `perf_context` [recently](https://github.com/pingcap/rust-rocksdb/pull/210). We can utilize `perf_context` to make it easy to diagnosis slow read issues related to tombstones.

This PR adds the ability to provide `perf_context` statistics for the Cursor. Later PRs will report statistics to the metrics in Coprocessor and Storage module.

- [x] Basic Impl
- [x] Test
- [ ] Benchmark (coming soon!)

## FAQ:

1. Why not integrate `PerfStatistics` into `CFStatistics`, for example, become one of its field?

   Note that the collector needs a `&mut PerfStatistics` in order to write statistics.

   So, if `PerfStatistics` is part of `CFStatistics`, there will be `&mut` violations: one is `&mut some_cf_statistics`, one is `&mut some_cf_statistics.perf_statistics`, which won't work.

2. Why `Cursor` accepts an optional `PerfStatistics`, i.e. `Option<PerfStatistics>`, unlike `CFStatistics`?

   Consider following scenario, `Cursor.get()`:

   ```
   Cursor.get
       Cursor.near_seek
           Cursor.seek
               Iterator.seek
           Cursor.next
               Iterator.next
       Cursor.near_seek_for_prev
       ...
   ```

   We record the count at the beginning of `Cursor.get()`, and record the count at the end of `Cursor.get()`, like this (pseudo code):

   ```
   Cursor.get
       perf_statistics = current_statistics();
       Cursor.near_seek
           Cursor.seek
               Iterator.seek
           Cursor.next
               Iterator.next
       Cursor.near_seek_for_prev
       ...
       delta_perf_statistics = current_statistics() - perf_statistics();
       add_perf_statistics(delta_perf_statistics);
   ```

   In this pattern, inner functions, like `Cursor.near_seek` should **NOT** contribute to perf_statistics again. Otherwise counters will be counted for multiple times.

   As you can see, naturally, we have two usages, one we collect perf statistics, one we don't.

3. Why not embed `PerfContextCollector` as a field of `Cursor`, and retrieve it for once after an operation?

   Good idea! But, `PerfContextCollector` is `!Send` && `!Sync`, so we cannot put it into a `Cursor: Send + Sync`, otherwise `Send + Sync` will break.

4. Why is `PerfContextCollector` not `Send` and not `Sync`?

   Because RocksDB's PerfContext is thread local.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

Unit test has been added to verify that we can actually get the perf values.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No. Metrics will not be changed in this PR.

## Benchmark result if necessary (optional)

Work in progress.

### CODE REVIEW

**Comment by breeswish** [comment]
@huachaohuang @zhangjinpeng1987 @hicqu PTAL

**Comment by huachaohuang** [comment]
OMG, this is too complicated. Why not just record the counters before handling a request and calculate the delta after finishing the request?

**Comment by breeswish** [comment]
@huachaohuang Hmm, that's approachable as well. Previously I wanted to avoid adding a collector in each request.

**Comment by breeswish** [comment]
@huachaohuang Seems that adding collector for each request entry should be simpler, since it is highly templated, which should be able to handle by macros. Can we refine and simplify our CFStatistics like this in future?

File: src/storage/mvcc/reader.rs [hunk 1]
```diff
@@ -60,6 +61,7 @@ impl MvccReader {
         MvccReader {
             snapshot,
             statistics: Statistics::default(),
+            perf_statistics: PerfStatistics::default(),
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: do we need to enable it by default?

File: src/storage/engine/perf_context.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    pub fn is_absolute(&self) -> bool {
+        self.absolute
+    }
+}
+
+/// Helper to collect RocksDB's `PerfContext`. This struct records the statistics at creation,
+/// and updates delta to the `collect_target` when it is dropped.
+#[must_use = "You must keep the collector live in the scope, otherwise it won't collect anything!"]
```
Comment (Hoverbear) [inline src/storage/engine/perf_context.rs]: `alive in the scope` :)

File: src/storage/engine/perf_context.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+/// RocksDB's `PerfContext` is thread local. If we send `PerfContextCollector` to another thread,
+/// we will get incorrect delta values, so it is `!Send` and `!Sync`.
+impl<'a> !Send for PerfContextCollector<'a> {}
+
+impl<'a> !Sync for PerfContextCollector<'a> {}
+
+/// A wrapper of `PerfContextCollector` to deal with optional targets. If the collecting target
+/// is `None`, there will be no overheads.
+pub enum OptionTargetPerfContextCollector<'a> {
```
Comment (Hoverbear) [inline src/storage/engine/perf_context.rs]: Hm, I'm not sure this is necessary. Maybe you can help me understand why this was added?

Comment (breeswish) [reply src/storage/engine/perf_context.rs]: See FAQ 2. Basically, for one function, we sometimes want it to collect metrics, but sometimes not. So this "OptionTargetPerfContextCollector" enables us to control whether or not to collect without much overhead.

**Comment by breeswish** [comment]
This implementation will be replaced by a simpler one. Closing.

**Pull request closed without merging**